File: py/play.py
# MODULES
import string
import re
import copy
import multiprocessing
import os
import math
from datetime import datetime


# IMPORTANT 'CONSTANTS'
LOGFILE = "log/progress_py.txt"
NUM_PROCESSES = os.cpu_count()
NUM_CHARACTERS = 5
ABC = [list(string.ascii_lowercase) for i in range(NUM_CHARACTERS)]
COLORS = {"green": 0, "yellow": 1, "grey": 2}
NUM_COMBOS = pow(len(COLORS), NUM_CHARACTERS)
COLOR_COMBOS = []
for a in COLORS:
    for b in COLORS:
        for c in COLORS:
            for d in COLORS:
                for e in COLORS:
                    COLOR_COMBOS.append([COLORS[a], COLORS[b], COLORS[c], COLORS[d], COLORS[e]])
WORDS = {}
with open("data/wordle_list.txt", "r") as file:
    for line in file:
        WORDS[line.strip()] = 0
with open("data/unigram_freq.csv", "r") as file:
    file.readline()  # read and discard the first line
    for line in file:
        word, freq = line.strip().split(",")
        freq = int(freq)
        if len(word) == NUM_CHARACTERS and word in WORDS:
            WORDS[word] = freq


# 'BUSINESS LOGIC' FUNCTIONS
# builds a regular expression to subset the word list to possible remaining words.
# NOTE: this function edits remaining_letters in-place.
def build_regex(guess, combo, remaining_letters):
    yellow_letters = []
    greys = set()
    nongreys = set()
    for i in range(NUM_CHARACTERS):
        current_letter = guess[i]
        current_combo_val = combo[i]
        # green letters are set
        if current_combo_val == COLORS["green"]:
            nongreys.add(current_letter)
            remaining_letters[i] = [current_letter]
        # yellow letters are removed from the index at which they are found
        elif current_combo_val == COLORS["yellow"]:
            nongreys.add(current_letter)
            yellow_letters.append(current_letter)
            remaining_letters[i].remove(current_letter)
        # grey letters are removed from each index
        else:
            greys.add(current_letter)
            for j in range(NUM_CHARACTERS):
                if current_letter in remaining_letters[j]:
                    remaining_letters[j].remove(current_letter)
    if len(greys & nongreys) > 0:
        raise ValueError("Letters cannot be both grey and nongrey!")
    rgx_components = []
    for i in range(NUM_CHARACTERS):
        if len(remaining_letters[i]) == 0:
            raise ValueError("Out of letters available for regex!")
        rgx_components.append("[" + "".join(remaining_letters[i]) + "]")
    rgx = re.compile(r"".join(rgx_components))
    return [rgx, yellow_letters]


# take the user's guess and filter down to the remaining
# possible words based on the input and color combo for that input.
def guess_filter(guess, combo, remaining_words, remaining_letters):
    subset = {}
    try:
        rgx, yellow_letters = build_regex(guess, combo, remaining_letters)
    except ValueError:
        return subset
    for word in remaining_words:
        if rgx.search(word):
            # ensure results contain each of the yellow letters when yellow was in the combo
            if len(yellow_letters) > 0:
                all_yellow_letters_found = True
                for yellow_letter in yellow_letters:
                    all_yellow_letters_found = all_yellow_letters_found and (yellow_letter in word)
                if all_yellow_letters_found:
                    subset[word] = remaining_words[word]
            else:
                subset[word] = remaining_words[word]
    return subset


# the parallel part of calculate_scores
def calculate_scores_(iterable, information):
    guesses, remaining_words, remaining_letters, freq_total = iterable
    for guess in guesses:
        formatted_time = datetime.now().strftime('%Y-%m-%d %H:%M:%S %Z')
        with open(LOGFILE, "a") as file:
            print(f"word: {guess}\ttime: {formatted_time}\n", end="", file=file)
        remaining_words_by_combo = []
        for i in range(len(COLOR_COMBOS)):
            filtered = guess_filter(guess,
                                    COLOR_COMBOS[i],
                                    remaining_words,
                                    copy.deepcopy(remaining_letters))
            if len(filtered) > 0:
                remaining_words_by_combo.append(filtered)
        # calculate expected bits of information gained
        expected_information = 0.0
        for i in range(len(remaining_words_by_combo)):
            combo_freq = sum(value for value in remaining_words_by_combo[i].values())
            proportion_of_words_remaining_for_this_combo = combo_freq / freq_total
            entropy = 0.0
            if proportion_of_words_remaining_for_this_combo > 0.0:
                entropy = math.log2(1.0 / proportion_of_words_remaining_for_this_combo)
            expected_information += proportion_of_words_remaining_for_this_combo * entropy
        information[guess] = expected_information


# calculate the bits of information gained for
# each guess after checking it against each color combination.
def calculate_scores(remaining_words, remaining_letters):
    freq_total = sum(value for value in remaining_words.values())
    words = list(remaining_words)
    numwords = len(words)
    words_each_process_is_responsible_for = []
    numwords_per_process = math.ceil(numwords / NUM_PROCESSES)
    start = 0
    for i in range(NUM_PROCESSES):
        words_each_process_is_responsible_for.append(words[start:(start + numwords_per_process)])
        start += numwords_per_process
        if(start + numwords_per_process > numwords):
            numwords_per_process = numwords_per_process - (numwords_per_process - numwords)
    iter = []
    for i in range(NUM_PROCESSES):
        iter.append([words_each_process_is_responsible_for[i],
                     remaining_words,
                     remaining_letters,
                     freq_total])
    if os.path.exists(LOGFILE):
        os.remove(LOGFILE)
    with multiprocessing.Pool(processes=NUM_PROCESSES) as pool:
        manager = multiprocessing.Manager()
        expected_information = manager.dict()
        pool.starmap(calculate_scores_,
                     [(elt, expected_information) for elt in iter]
                     )
    return dict(expected_information)


# UNIT TESTS
# this answer matches the build_regex call in play.R (after adding the re.compile call)
build_regex("helio", [1, 2, 2, 0, 1], copy.deepcopy(ABC))[0] == re.compile("[abcdfgijkmnopqrstuvwxyz][abcdfghijkmnopqrstuvwxyz][abcdfghijkmnopqrstuvwxyz][i][abcdfghijkmnpqrstuvwxyz]")
# these answers match the guess_filter call in play.R
"".join(guess_filter("ocean", [0, 0, 0, 0, 0], copy.deepcopy(WORDS), copy.deepcopy(ABC)).keys()) == "ocean"
"".join(guess_filter("ocean", [0, 0, 2, 0, 0], copy.deepcopy(WORDS), copy.deepcopy(ABC)).keys()) == "octan"
len(guess_filter("ocean", [0, 0, 1, 0, 0], copy.deepcopy(WORDS), copy.deepcopy(ABC))) == 0
# these answers match the calculate_scores call in play.R
testdict = {"aahed": WORDS["aahed"],
            "aalii": WORDS["aalii"],
            "aargh": WORDS["aargh"],
            "aarti": WORDS["aarti"],
            "abaca": WORDS["abaca"]}
outdict = calculate_scores(testdict, ABC)
round(outdict["aahed"], 1) == 1.6
round(outdict["aalii"], 1) == 1.6
round(outdict["aargh"], 1) == 1.6
round(outdict["aarti"], 1) == 1.6
round(outdict["abaca"], 1) == 1.0


# EXECUTION
scores = calculate_scores(WORDS, ABC)
outfile = "data/opening_word_scores.tsv"
if os.path.exists(outfile):
    os.remove(outfile)
with open(outfile, "a") as file:
    print("word\texpected_entropy\tfrequency", file=file)
    for word in scores:
        print(f"{word}\t{scores[word]}\t{WORDS[word]}", file=file)

File: jl/play.jl
using Base.Threads
using Dates # only used for logging

# IMPORTANT 'CONSTANTS'
const UTC_TO_LOCAL_DIFF = -(60 * 60 * 5)
const TZ = "EST"
const NUM_PROCESSES = Threads.nthreads()
const NUM_CHARACTERS = 5
const COLORS = Dict(:green => 0, :yellow => 1, :grey => 2)
const LETTERS = ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z"]
abc = Vector{Array}(undef, NUM_CHARACTERS)
for i in 1:NUM_CHARACTERS
    abc[i] = LETTERS
end
const ABC = abc
const NUM_COMBOS = length(COLORS)^NUM_CHARACTERS
color_combos = Vector{Array}(undef, NUM_COMBOS)
global counter = 0
for i in keys(COLORS)
    for ii in keys(COLORS)
        for iii in keys(COLORS)
            for iv in keys(COLORS)
                for v in keys(COLORS)
                    global counter += 1
                    color_combos[counter] = [COLORS[i], COLORS[ii], COLORS[iii], COLORS[iv], COLORS[v]]
                end
            end
        end
    end
end
const COLOR_COMBOS = color_combos
words = open(io->read(io, String), "data/wordle_list.txt")
words = words[1:(length(words) - 1)] # remove trailing newline
words = string.(split(words, "\n"))
words = Dict{String, Int}(word => 0 for word in words)
unigrams = open(io->read(io, String), "data/unigram_freq.csv")
unigrams = split(unigrams, "\n")
unigrams = split.(unigrams, ",")[2:(length(unigrams) - 1)] # remove header, trailing newline
filter!(row->length(row[1]) == 5, unigrams)
for (word, freq) in unigrams
    if(word in keys(words))
        words[word] = parse(Int, freq)
    end
end
const WORDS = words
tz = nothing
abc = nothing
counter = nothing
color_combos = nothing
words = nothing
unigrams = nothing

# HELPER FUNCTIONS
collapse_into_character_groups = function(string_arr)
    "[" * join(string_arr) * "]"
end
generate_logfile_name = function(id)
    "log/progress_jl" * string(id) * ".txt"
end
format_time = function()
    formatted_datetime = Dates.unix2datetime(time() + UTC_TO_LOCAL_DIFF)
    Dates.format(formatted_datetime, "yyyy-mm-dd HH:MM:SS") * " " * TZ
end
print_log_info = function(path_logfile, guess, starttime)
    info = "word: " * guess * "\tstart: " * starttime * "\tend: " * format_time()
    open(path_logfile, "a+") do logfile
        println(logfile, info)
    end
end

# 'BUSINESS LOGIC' FUNCTIONS
# builds a regular expression to subset the word list to possible remaining words.
# NOTE: this function edits remaining_letters in-place.
build_regex = function(guess, combo, remaining_letters)
    yellow_letters = []
    greys = Set()
    nongreys = Set()
    for i in 1:NUM_CHARACTERS
        current_letter = string(guess[i])
        current_combo_val = combo[i]
        # green letters are set
        if current_combo_val == COLORS[:green]
            push!(nongreys, current_letter)
            remaining_letters[i] = [string(current_letter)]
        # yellow letters are removed from the index at which they are found
        elseif current_combo_val == COLORS[:yellow]
            push!(nongreys, current_letter)
            push!(yellow_letters, current_letter)
            filter!(letter->letter != current_letter, remaining_letters[i])
        # grey letters are removed from each index
        else
            push!(greys, current_letter)
            filter!.(letter->letter != current_letter, remaining_letters)
        end
    end
    # this is a fine check to have, but the color combos are already checked upstream
    if any(grey -> grey ∈ nongreys, greys)
        error("Letters cannot be both grey and nongrey!")
    end
    if any(letter_arr -> length(letter_arr) == 0, remaining_letters)
        error("Out of letters available for regex!")
    end
    rgx = Regex(join(collapse_into_character_groups.(remaining_letters)))
    [rgx, yellow_letters]
end
# take the user's guess and filter down to the remaining
# possible words based on the input and color combo for that input.
guess_filter = function(guess, combo, remaining_words, remaining_letters)
    subset = Dict{String, Int}()
    rgx = nothing
    yellow_letters = nothing
    try
        rgx, yellow_letters = build_regex(guess, combo, remaining_letters)
    catch
        return subset
    end
    for word in keys(remaining_words)
        if contains(word, rgx)
            if all(occursin.(yellow_letters, word))
                subset[word] = remaining_words[word]
            end
        end
    end
    subset
end
# calculate the bits of information gained for
# each guess after checking it against each color combination.
calculate_scores = function(remaining_words, remaining_letters, consolidate_logsp)
    # prepare iterator
    freq_total = sum(values(remaining_words))
    words = string.(keys(remaining_words))
    numwords = length(words)
    if numwords < NUM_PROCESSES
        numprocs = 1
    else
        numprocs = NUM_PROCESSES
    end
    words_each_process_is_responsible_for = Vector{Array}(undef, numprocs)
    numwords_per_process = Int(ceil(numwords / numprocs))
    start = 1
    for i in 1:numprocs
        words_each_process_is_responsible_for[i] = words[start:(start + numwords_per_process - 1)]
        start += numwords_per_process
        if(start + numwords_per_process > numwords)
            numwords_per_process = numwords - (start - 1)
        end
    end

    # compute scores
    expected_information = Dict{String, Float64}()
    Threads.@threads for iter in 1:numprocs
        guesses = words_each_process_is_responsible_for[iter]
        logfile = generate_logfile_name(iter)
        open(logfile, "w") # create log
        for guess in guesses
            guess_start_time = format_time()
            remaining_words_by_combo = Vector(undef, NUM_COMBOS)
            for i in eachindex(COLOR_COMBOS)
                filtered = keys(guess_filter(guess, COLOR_COMBOS[i], remaining_words, deepcopy.(remaining_letters)))
                if(length(filtered) > 0)
                    remaining_words_by_combo[i] = filtered
                else
                    remaining_words_by_combo[i] = []
                end
            end
            remaining_words_by_combo = remaining_words_by_combo[length.(remaining_words_by_combo) .> 0]

            # calculate expected bits of information gained
            info = 0
            for i in eachindex(remaining_words_by_combo)
                combo_freq = 0
                for remaining_word in remaining_words_by_combo[i]
                    combo_freq += sum(remaining_words[remaining_word])
                end
                proportion_of_words_remaining_for_this_combo = combo_freq / freq_total
                entropy = ifelse(proportion_of_words_remaining_for_this_combo > 0,
                                 log2(1 / proportion_of_words_remaining_for_this_combo),
                                 0)
                info += proportion_of_words_remaining_for_this_combo * entropy
            end
            expected_information[guess] = info
            print_log_info(logfile, guess, guess_start_time)
        end
    end

    # consolidate logs
    outfile = generate_logfile_name("")
    logfiles = generate_logfile_name.(1:numprocs)
    if(consolidate_logsp)
        open(outfile, "w") do outfile
            for logfile in logfiles
                open(logfile, "r") do logfiles
                    write(outfile, read(logfile, String))
                end
            end
        end
    end
    rm.(logfiles)

    expected_information
end

# EXECUTION
scores = calculate_scores(WORDS, ABC, true)
scores_file = "data/opening_word_scores.tsv"
open(scores_file, "w") do scores_file
    write(scores_file, "word\texpected_entropy\tfrequency\n")
    for (k,v) in scores
        line = string(k) * "\t" * string(v) * "\t" * string(WORDS[k]) * "\n"
        write(scores_file, line)
    end
end
